server/channels/jobs/batch_worker.go

```diff
@@ -73,10 +84,14 @@ func (worker *BatchWorker) Run() {
 
 // Stop interrupts the worker even if the migration has not yet completed.
 func (worker *BatchWorker) Stop() {
+	worker.stateMut.Lock()
+	defer worker.stateMut.Unlock()
+
 	// Set to close, and if already closed before, then return.
-	if !worker.closed.CompareAndSwap(false, true) {
+	if worker.closed {
 		return
 	}
+	worker.closed = true
 
 	worker.logger.Debug("Worker stopping")
 	close(worker.stop)
```

Comment:
Do we need to hold onto the lock while we wait for the worker to stop?

Reply:
Yeah I'd say it's safer and more robust to do that.

Reply:
Ah, so I was wondering about exactly the reverse: if we're holding onto a lock but explicitly blocked on another process, there's a non-zero chance that deadlock with that other process who might also need the lock. Unless there's something we need to guard here, I'm wondering if it's safer to just release the lock when we're genuinely done.

Reply:
Ah, I see my mistake now: there's no further blocking calls in this method. I'm not sure where I thought I saw one earlier... sorry for the thrash :D